src/helpers/helpers.ts

```diff
@@ -1,59 +1,100 @@
 import { wikibaseTimeToDateObject as toDateObject } from './wikibase_time_to_date_object.js'
+import type {
+  EntityId,
+  EntityPageTitle,
+  EntitySchemaId,
+  FormId,
+  Guid,
+  Hash,
+  ItemId,
+  LexemeId,
+  NonNestedEntityId,
+  NumericId,
+  PropertyClaimsId,
+  PropertyId,
+  RevisionId,
+  SenseId,
+} from '../types/entity.js'
 import type { Url } from '../types/options.js'
 
-export const isNumericId = (id: string): boolean => /^[1-9][0-9]*$/.test(id)
+export function isNumericId (id: unknown): id is NumericId {
+  return typeof id === 'string' && /^[1-9][0-9]*$/.test(id)
+}
 
-export const isEntityId = (id: string): boolean => /^((Q|P|L|M)[1-9][0-9]*|L[1-9][0-9]*-(F|S)[1-9][0-9]*)$/.test(id)
+export function isEntityId (id: unknown): id is EntityId {
+  return typeof id === 'string' &&
+    /^((Q|P|L|M)[1-9][0-9]*|L[1-9][0-9]*-(F|S)[1-9][0-9]*)$/.test(id)
+}
```

Comment:
In pure JS, I would have been tempted to use a factory function, such as
```js
function TypeTest (pattern) {
  return function (id) {
    return typeof id === 'string' && pattern.test(id)
  }
}
const entityPattern = /^((Q|P|L|M)[1-9][0-9]*|L[1-9][0-9]*-(F|S)[1-9][0-9]*)$/
export const isEntityId = TypeTest(entityPattern)
```
but in TS, that seems less of a good idea: is there a way in TS to use such a factory function, while keeping the typing?

Reply:
First of all: im not even sure if this factory function might be dangerous in the first place as RegEx can have side effects in JS.
I don't think these RegEx have side effects as mainly the g(lobal) specifier results in side effects and state within the RegEx object but still. RegEx is able to have state and side effects so I would try to prevent that.

In general, you could type such builder functions:

```ts
function builder<T extends string>(
  regex: {readonly source: string},
): (id: string) => id is T {
  // this should assume the types from the return type of the builder function
  return id => {
    return typeof id === 'string' && new RegExp(regex.source).test(id)
  }
}

const entityPattern = /^((Q|P|L|M)[1-9][0-9]*|L[1-9][0-9]*-(F|S)[1-9][0-9]*)$/
export const isEntityId = builder<EntityId>(entityPattern)
```

This one creates a new RegExp from the source of the original one to prevent that RegExp internal state dilemma I talked about.

(Note: This is purely written in a browser an not tested with any kind of TypeScript compiler. It might not work but assume it does)

Reply:
this is quite nice, I added it in 68d6d63